### DISCUSSION TITLE
Allow for Forward search in explore Loki

### DISCUSSION THREAD

**Discussion by cyriltovena** [grafana/grafana#56850] [Observability]
**What would you like to be added**:

I want an option to search forward within logs currently we can only do backward.

**Why is this needed**:

Backward is nice when you're looking at the last logs of an applications but forward is more useful when you're scanning past incidents. 

It would be also nice to be able to page from the last logs received so we can move within the same timeframe but show missing logs cut by the limit.

Loki API support both forward and backward search.

**Comment by jonathancrone** [comment]
Agreed: 
Example:  you end up with a time-based crash, and are trying to confirm whether the crash is the same behaviour each time through:   forward or backwards is "normal" search behaviour.    Going "old school" for a moment, there's a reason why a "pager"  like 'less'  has both / and ? for searching in a file...

**Comment by midnightconman** [comment]
This would also help with some of the lower line limits set by default in loki and grafana... most times I search in grafana (using loki datasource) for a pattern (even over an hour or less), the first 10 minutes (t-60m to t-50m) of the search return results and then the limit is hit. It would be nice to have the option to also search for newest lines first (t-0m).

**Comment by mizeng** [comment]
@davkal  I see the feature in #16555, this is only let frontend to revise the output with still uses "direction=Backward". Could we support UI level switch to choose calling "direction=Backward" or "direction=Forward"? 

I think it's reasonable that some people perfer to see logs in natural time order. Why not give them an option in Grafana UI?

**Comment by jbarotin** [comment]
Hi,
> Could we support UI level switch to choose calling "direction=Backward" or "direction=Forward"?

This kind of improvement will be great. Any chance to get it implemented soon?

**Comment by brunokoeferli** [comment]
Hi,
are there any news/plans regarding this topic? As noted in https://github.com/grafana/grafana/pull/43860, are UX discussions still ongoing?

We are considering switching from Kibana/Elasticsearch to Grafana/Loki. So far it was sometimes quite useful to browse logs in `direction=forward`.

**Comment by jtackaberry** [comment]
IMO it's very confusing that setting the time range to "Last 12 hours" and clicking "Oldest first" for the results order actually produces logs from the most recent minute.  When I demoed Loki to a group of potential users, their intuition was that this should have conducted a forward search and the first batch of results should have been from 12 hours ago.

IMO choosing "oldest first" should implicitly perform a forward search from the start of the time range, rather than a backward search from the end of the time range and just sorting the results in ascending order.

**Comment by jkemp101** [comment]
I've been playing with this a bit and not completely clear what is and is not supported. It looks like there is some[ backend support](https://github.com/search?q=repo%3Agrafana%2Fgrafana+LokiQueryDirection&type=code) to search Loki in a forward direction. Manually playing around with it by putting `"direction":"forward"` in the URL has given me mixed results. It works for one query but another one only returns 2 lines. Maybe it busts the cache and doesn't know how to switch between forward and backward for the same query? Is there a GH Issue or discussion that clarifies want is left to do so that we can search logs from older to newest, not just sort the current set in the UI? Am I missing something in the UI that lets you send the `"direction":"forward"` to the backend?

**Comment by bygui86** [comment]
any news about this?!

**Comment by tudburley** [comment]
This is actually a blocker for us to migrate from ELK (Kibana) to Loki. Any news about this?

**Comment by ivanahuckova** [comment]
For additional discussion, please use https://github.com/grafana/grafana/issues/78588.
